experimental/report/writer.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"bytes"
+	"io"
+	"strings"
+	"unicode"
+
+	"github.com/bufbuild/protocompile/internal/ext/unsafex"
+)
+
+// writer implements low-level writing helpers, including a custom buffering
+// routine to avoid printing trailing whitespace to the output.
+type writer struct {
+	out io.Writer
+	buf []byte
+}
+
+// Write implements [io.Writer].
+func (w *writer) Write(data []byte) (int, error) {
+	w.WriteString(unsafex.StringAlias(data))
+	return len(data), nil
+}
+
+func (w *writer) WriteBytes(b byte, n int) {
```

Comment:
I think passing `b = '\n'` breaks this thing. This is only ever called with a space. How about instead just name it `WritePadding(n int)` and hardcode the space?

Reply:
Don't think `b = '\n'` breaks as the same `WriteString("\n\n\n")` would append `\n\n\n` to the buffer, just looping over and checking for right padding between each newline.

Reply:
No, Josh is right. Plus, having it only print spaces means we can make it more efficient.